lib/icinga/clusterevents-check.cpp

```diff
@@ -93,13 +93,93 @@ void ClusterEvents::ExecuteCheckFromQueue(const MessageOrigin::Ptr& origin, cons
 
 	if (!listener) {
-		Log(LogCritical, "ApiListener", "No instance available.");
+		Log(LogCritical, "ApiListener") << "No instance available.";
 		return;
 	}
 
+	Defer resetExecuteCommandProcessFinishedHandler ([]() {
+		Checkable::ExecuteCommandProcessFinishedHandler = nullptr;
+	});
+
 	if (params->Contains("source")) {
-		Log(LogCritical, "ApiListener", "Not implemented.");
-		return;
+		String uuid = params->Get("source");
+
+		Host::Ptr host = Host::GetByName(params->Get("host"));
+		if (!host) {
+			Log(LogCritical, "ApiListener") << "Host '" << params->Get("host") << "' not found.";
+			return;
+		}
+
+		Checkable::Ptr checkable;
+		if (params->Contains("service"))
+			checkable = host->GetServiceByShortName(params->Get("service"));
+		else
+			checkable = host;
+
+		if (!checkable) {
+			Log(LogCritical, "ApiListener") << "Checkable '" << params->Get("host")
+				<< "!" << params->Get("service") << "' not found.";
+			return;
+		}
+
+		ObjectLock oLock (checkable);
+
+		if (origin->FromZone && !origin->FromZone->CanAccessObject(checkable)) {
+			Log(LogNotice, "ApiListener")
+				<< "Discarding 'ExecuteCheckFromQueue' event for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': Unauthorized access.";
+			return;
+		}
+
+		/* Check deadline */
+		double deadline = params->Get("deadline");
+		if (Utility::GetTime() > deadline) {
+			Log(LogNotice, "ApiListener")
+				<< "Discarding 'ExecuteCheckFromQueue' event for checkable '" << checkable->GetName()
+				<< "' from '" << origin->FromClient->GetIdentity() << "': Deadline has expired.";
+			return;
+		}
+
+		Checkable::ExecuteCommandProcessFinishedHandler = [checkable, listener, sourceEndpoint, origin, params] (const Value& commandLine, const ProcessResult& pr) {
+			Checkable::CurrentConcurrentChecks.fetch_sub(1);
+			Checkable::DecreasePendingChecks();
+
+			if (pr.ExitStatus > 3) {
+				Process::Arguments parguments = Process::PrepareCommand(commandLine);
+				Log(LogWarning, "ApiListener")
+					<< "Check command for object '" << checkable->GetName() << "' (PID: " << pr.PID
+					<< ", arguments: " << Process::PrettyPrintArguments(parguments) << ") terminated with exit code "
+					<< pr.ExitStatus << ", output: " << pr.Output;
+			}
+
+			String output = pr.Output.Trim();
+
+			std::pair<String, String> co = PluginUtility::ParseCheckOutput(output);
+			CheckResult::Ptr cr = new CheckResult();
+			cr->SetCommand(commandLine);
+			cr->SetOutput(co.first);
+			cr->SetPerformanceData(PluginUtility::SplitPerfdata(co.second));
+			cr->SetState(PluginUtility::ExitStatusToState(pr.ExitStatus));
+			cr->SetExitStatus(pr.ExitStatus);
+			cr->SetExecutionStart(pr.ExecutionStart);
+			cr->SetExecutionEnd(pr.ExecutionEnd);
+
+			Dictionary::Ptr executedParams = new Dictionary();
+			params->CopyTo(executedParams);
+			executedParams->Set("execution", params->Get("source"));
+			executedParams->Set("check_result", Serialize(cr));
```

Comment:
Please [have a look](https://github.com/Icinga/icinga2/pull/8040#issuecomment-638088361) which `executedParams` we need here.

Reply:
You're still sending a serialized CheckResult, not the params from [here](https://github.com/Icinga/icinga2/pull/8040#issuecomment-638088361).